From e05bae6e708f92b6ba0338652dfe4ddb51bfb9ad Mon Sep 17 00:00:00 2001
From: Li Wang <liwang@redhat.com>
Date: Thu, 20 Jun 2024 10:16:35 +0800
Subject: [PATCH] tst_kconfig: Avoid reporting buffer overflow when parsing
 /proc/cmdline

When the test is run with a kernel booting with many parameters, the
buffer size is often not large enough to store the complete command
line. This results in a buffer overflow and the test complains with
the following message:

  tst_kconfig.c:609: TWARN: Buffer overflowed while parsing /proc/cmdline

Note:

Petr point out that these configs, which are generated by toolchain will
be longer than 128 chars someday, but I don't think that is the reason
we need raise our parsed buffer, since tst_kcmdline_parse() was just add
for popular parameter (which always pass by user and short). So far I
don't see any LTP test parse a longer parameters.

Fixes: 180834982 ("kconfig: add funtion to parse /proc/cmdline")
Signed-off-by: Li Wang <liwang@redhat.com>
Reviewed-by: Petr Vorel <pvorel@suse.cz>
---
 include/tst_kconfig.h | 2 +-
 lib/tst_kconfig.c     | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/include/tst_kconfig.h b/include/tst_kconfig.h
index dcb370574..23f807409 100644
--- a/include/tst_kconfig.h
+++ b/include/tst_kconfig.h
@@ -87,7 +87,7 @@ int tst_kconfig_check(const char *const kconfigs[]);
  */
 struct tst_kcmdline_var {
 	const char *key;
-	char value[128];
+	char value[256];
 	bool found;
 };
 
diff --git a/lib/tst_kconfig.c b/lib/tst_kconfig.c
index e16ca1400..6d6b1da18 100644
--- a/lib/tst_kconfig.c
+++ b/lib/tst_kconfig.c
@@ -569,7 +569,7 @@ char tst_kconfig_get(const char *confname)
 
 void tst_kcmdline_parse(struct tst_kcmdline_var params[], size_t params_len)
 {
-	char buf[128], line[512];
+	char buf[256], line[1024];
 	size_t b_pos = 0,l_pos =0, i;
 	int var_id = -1;
 
@@ -606,7 +606,7 @@ void tst_kcmdline_parse(struct tst_kcmdline_var params[], size_t params_len)
 		break;
 		default:
 			if (b_pos + 1 >= sizeof(buf)) {
-				tst_res(TWARN, "Buffer overflowed while parsing /proc/cmdline");
+				tst_res(TINFO, "WARNING: Buffer overflowed while parsing /proc/cmdline");
 				while (line[l_pos] != '\0' && line[l_pos] != ' ' && line[l_pos] != '\n')
 					l_pos++;
 
-- 
2.45.2

